server/analyzer/serial.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"fmt"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/analyzer"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+
+	"github.com/dolthub/doltgresql/core"
+	"github.com/dolthub/doltgresql/core/sequences"
+	pgexprs "github.com/dolthub/doltgresql/server/expression"
+	"github.com/dolthub/doltgresql/server/functions/framework"
+	pgnodes "github.com/dolthub/doltgresql/server/node"
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// ReplaceSerial replaces a CreateTable node containing a SERIAL type with a node that can create sequences alongside
+// the table.
+func ReplaceSerial(ctx *sql.Context, a *analyzer.Analyzer, node sql.Node, scope *plan.Scope, selector analyzer.RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	createTable, ok1 := node.(*plan.CreateTable)
+	if !ok1 {
+		return node, transform.SameTree, nil
+	}
+
+	var ctSequences []*pgnodes.CreateSequence
+	for _, col := range createTable.PkSchema().Schema {
+		if doltgresType, ok2 := col.Type.(pgtypes.DoltgresType); ok2 {
+			isSerial := false
+			var maxValue int64
+			switch doltgresType.BaseID() {
+			case pgtypes.DoltgresTypeBaseID_Int16Serial:
+				isSerial = true
+				col.Type = pgtypes.Int16
+				maxValue = 32767
+			case pgtypes.DoltgresTypeBaseID_Int32Serial:
+				isSerial = true
+				col.Type = pgtypes.Int32
+				maxValue = 2147483647
+			case pgtypes.DoltgresTypeBaseID_Int64Serial:
+				isSerial = true
+				col.Type = pgtypes.Int64
+				maxValue = 9223372036854775807
+			}
+			if isSerial {
+				baseSequenceName := fmt.Sprintf("%s_%s_seq", createTable.Name(), col.Name)
+				sequenceName := baseSequenceName
+				relationType, err := core.HasRelation(ctx, "", baseSequenceName)
```

Comment:
Probably want a TODO here to get the schema name from the CreateTable node

Better yet, a skipped test that requires this behavior